File: samples/client/3_0_3_unit_test/java/src/test/java/org/openapijsonschematools/client/components/schemas/AllofWithBaseSchemaTest.java
package org.openapijsonschematools.client.components.schemas;

import org.junit.Assert;
import org.junit.Test;
import org.openapijsonschematools.client.configurations.JsonSchemaKeywordFlags;
import org.openapijsonschematools.client.configurations.SchemaConfiguration;
import org.openapijsonschematools.client.exceptions.ValidationException;
import org.openapijsonschematools.client.schemas.MapMaker;
import org.openapijsonschematools.client.schemas.validation.JsonSchema;
import org.openapijsonschematools.client.schemas.validation.FrozenMap;
import org.openapijsonschematools.client.schemas.validation.FrozenList;
import org.openapijsonschematools.client.schemas.validation.PathToSchemasMap;
import org.openapijsonschematools.client.schemas.validation.ValidationMetadata;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.AbstractMap;
import java.util.LinkedHashSet;

public class AllofWithBaseSchemaTest {
    static final SchemaConfiguration configuration = new SchemaConfiguration(JsonSchemaKeywordFlags.ofNone());
    static final ValidationMetadata validationMetadata = new ValidationMetadata(
            List.of("args[0"),
            configuration,
            new PathToSchemasMap(),
            new LinkedHashSet<>()
    );

    @Test
    public void testMismatchBaseSchemaFails() {
        // mismatch base schema
        final var schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<String, @Nullable String>(
                    "foo",
                    "quux"
                ),
                new AbstractMap.SimpleEntry<String, @Nullable String>(
                    "baz",
                    null
                )
            ),
            validationMetadata
        ));
    }

    @Test
    public void testMismatchFirstAllofFails() {
        // mismatch first allOf
        final var schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<String, @Nullable Long>(
                    "bar",
                    2L
                ),
                new AbstractMap.SimpleEntry<String, @Nullable Long>(
                    "baz",
                    null
                )
            ),
            validationMetadata
        ));
    }

    @Test
    public void testValidPasses() {
        // valid
        final var schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance();
        schema.validate(
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<String, @Nullable Object>(
                    "foo",
                    "quux"
                ),
                new AbstractMap.SimpleEntry<String, @Nullable Object>(
                    "bar",
                    2L
                ),
                new AbstractMap.SimpleEntry<String, @Nullable Object>(
                    "baz",
                    null
                )
            ),
            configuration
        );
    }

    @Test
    public void testMismatchBothFails() {
        // mismatch both
        final var schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<String, Long>(
                    "bar",
                    2L
                )
            ),
            validationMetadata
        ));
    }

    @Test
    public void testMismatchSecondAllofFails() {
        // mismatch second allOf
        final var schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance();
        Assert.assertThrows(ValidationException.class, () -> JsonSchema.validate(
            schema,
            MapMaker.makeMap(
                new AbstractMap.SimpleEntry<String, Object>(
                    "foo",
                    "quux"
                ),
                new AbstractMap.SimpleEntry<String, Object>(
                    "bar",
                    2L
                )
            ),
            validationMetadata
        ));
    }
}

File: samples/client/3_0_3_unit_test/kotlin/src/test/kotlin/org/openapijsonschematools/client/components/schemas/AllofWithBaseSchemaTest.kt
package org.openapijsonschematools.client.components.schemas

import org.openapijsonschematools.client.configurations.JsonSchemaKeyword
import org.openapijsonschematools.client.configurations.SchemaConfiguration
import org.openapijsonschematools.client.exceptions.ValidationException

import kotlin.test.Test
import kotlin.test.assertFailsWith

class AllofWithBaseSchemaTest {
    companion object {
        val configuration = SchemaConfiguration(disabledKeywordFlags=setOf(JsonSchemaKeyword.FORMAT))
    }

    @Test
    fun testMismatchBaseSchemaFails() {
        // mismatch base schema
        val schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "foo",
                            "quux"
                        ),
                        Pair(
                            "baz",
                            null
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testMismatchFirstAllofFails() {
        // mismatch first allOf
        val schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "bar",
                            2
                        ),
                        Pair(
                            "baz",
                            null
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testValidPasses() {
        // valid
        val schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance()
        schema.validate(
            mapOf(
                Pair(
                    "foo",
                    "quux"
                ),
                Pair(
                    "bar",
                    2
                ),
                Pair(
                    "baz",
                    null
                )
            ),
            configuration
        )
    }

    @Test
    fun testMismatchBothFails() {
        // mismatch both
        val schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "bar",
                            2
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testMismatchSecondAllofFails() {
        // mismatch second allOf
        val schema = AllofWithBaseSchema.AllofWithBaseSchema1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "foo",
                            "quux"
                        ),
                        Pair(
                            "bar",
                            2
                        )
                    ),
                    configuration
                )
            }
        )
    }
}
